server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -109,13 +115,24 @@ public void onNewInfo(ClusterInfo info) {
             }
             ClusterState state = clusterStateSupplier.get();
             Set<String> indicesToMarkReadOnly = new HashSet<>();
+            RoutingNodes routingNodes = state.getRoutingNodes();
+            Map<String, Boolean> indexAutoReleaseEligibility = new HashMap<>();
+            // Ensure we release indices on nodes that have a usage response from node stats
+            markNodesMissingUsageIneligibleForRelease(routingNodes, usages, indexAutoReleaseEligibility);
             for (ObjectObjectCursor<String, DiskUsage> entry : usages) {
                 String node = entry.key;
                 DiskUsage usage = entry.value;
                 warnAboutDiskIfNeeded(usage);
+                RoutingNode routingNode = state.getRoutingNodes().node(node);
+                // Only unblock index if all nodes that contain shards of it are below the high disk watermark
+                if (usage.getFreeBytes() < diskThresholdSettings.getFreeBytesThresholdHigh().getBytes()
+                    || usage.getFreeDiskAsPercentage() < diskThresholdSettings.getFreeDiskThresholdHigh()) {
```

Comment:
>I am not quite sure if we should have the 5% below the flood watermark check for auto-release since in most of the cases threshold below the high water should be good enough.
5% may just be too large for bigger instances(>1Tb). Similarly 5gb might also be too big for smaller instances (<10Gb). In which case customers might need a new setting to get auto-release at par with their existing watermark settings.
IMHO I think a new setting here would only complicate stuffs and most of the times auto-release below the high watermark might suffice(assuming customer has configured them correctly). Setting the watermark very close or configuring the same value for low/high/flood defeats the purpose

Quoting the comment from the issue

Reply:
I think it's important to offer some protection against flapping if the `high` and `flood_stage` watermarks are set equal. I have marked this issue for further team discussion.

Reply:
Ok, we discussed this some more and concluded that it wouldn't be _so_ bad if we just used the `high` watermark, even if it's set to equal the `flood_stage` one.

Reply:
Thanks @DaveCTurner will be great if you could review the new revision along with the tests.